nowcasting_dataset/data_sources/optical_flow/optical_flow_data_source.py

```diff
@@ -0,0 +1,267 @@
+            satellite_data: Satellite DataArray
+            t0_dt: t0 timestamp
+
+        Returns:
+            The Tensor with the optical flow predictions for t0 to forecast horizon
+        """
+
+        # Get the previous timestamp
+        future_timesteps = self._get_number_future_timesteps(satellite_data, t0_dt)
+        satellite_data: xr.DataArray = self._get_previous_timesteps(
+            satellite_data,
+            t0_dt=t0_dt,
+        )
+        prediction_block = np.zeros(
+            (
+                future_timesteps,
+                self.image_size_pixels,
+                self.image_size_pixels,
+                satellite_data.sizes["channels_index"],
+            )
+        )
+        for prediction_timestep in range(future_timesteps):
+            for channel in range(0, len(satellite_data.coords["channels_index"]), 4):
+                # Optical Flow works with RGB images, so chunking channels for it to be faster
+                channel_images = satellite_data.sel(channels_index=slice(channel, channel + 3))
+                # Extra 1 in shape from time dimension, so removing that dimension
+                t0_image = channel_images.isel(
+                    time_index=len(satellite_data.time_index) - 1
+                ).data.values
+                previous_image = channel_images.isel(
+                    time_index=len(satellite_data.time_index) - 2
+                ).data.values
+                optical_flow = self._compute_optical_flow(t0_image, previous_image)
+                # Do predictions now
+                flow = (
+                    optical_flow * prediction_timestep + 1
+                )  # Otherwise first prediction would be 0
+                warped_image = self._remap_image(t0_image, flow)
+                warped_image = self._crop_center(
+                    warped_image,
+                    self.image_size_pixels,
+                    self.image_size_pixels,
+                )
+                prediction_block[prediction_timestep, :, :, channel : channel + 4] = warped_image
+        dataarray = self._update_dataarray_with_predictions(
+            satellite_data=self._data, predictions=prediction_block
+        )
+        return dataarray
+
+    def _compute_optical_flow(self, t0_image: np.ndarray, previous_image: np.ndarray) -> np.ndarray:
+        """
+        Compute the optical flow for a set of images
+
+        Args:
+            t0_image: t0 image
+            previous_image: previous image to compute optical flow with
+
+        Returns:
+            Optical Flow field
+        """
+        # Input images have to be single channel and between 0 and 1
+        image_min = np.min([t0_image, previous_image])
+        image_max = np.max([t0_image, previous_image])
+        t0_image -= image_min
+        t0_image /= image_max
+        previous_image -= image_min
+        previous_image /= image_max
+        t0_image = cv2.cvtColor(t0_image.astype(np.float32), cv2.COLOR_RGBA2GRAY)
```

Comment:
Ah, I see - I was curious how the code operated on 4 channels at once!

My guess is that we should perform optical flow on each satellite channel individually, instead of converting 4 channels to grayscale and doing optical flow on the grayscale.

Converting "true" RGB images to grayscale kind of makes sense.  But, of course, satellite channels aren't RGB.  For example, OpenCV uses different weightings for red, green and blue (see [the OpenCV docs here](https://docs.opencv.org/4.5.4/de/d25/imgproc_color_conversions.html)) which makes sense for "true" RGB because we perceive red, green and blue differently.  OpenCV's RGB to grayscale conversion is optimised to look _perceptually_ good to humans; not to maximise the preservation of information :slightly_smiling_face: 

Unlike "true" RGB images, the different satellite channels might be moving in different directions.   e.g. the water vapour channel (which I think sees the entire depth of the atmosphere) might be drifting in one direction, whilst the top of the clouds (as seen by the IR channel) might be drifting in another direction.

So, yeah, my hunch would be that we need to work on individual satellite channels, one-at-a-time (after rescaling to [0, 255] and converting to `uint8`).  Which I appreciate will be slower!

That said,  if you're certain that RGB to grayscale conversion works for satellite images, then:

* I'd definitely do _RGB_ to grayscale, not RGB__A__ to grayscale.  The alpha channel is an opacity mask!
* I could be wrong but I think float32 images input into `cvtColor` need to be between 0 and 1.  (See the OpenCV docs [here](https://docs.opencv.org/4.5.4/d8/d01/group__imgproc__color__conversions.html#ga397ae87e1288a81d2363b61574eb8cab))

Reply:
Yeah, doing it individually makes sense, just changed it to doing it individually! I missed this bit on how OpenCV uses different weightings, so yeah, good idea to not do it.